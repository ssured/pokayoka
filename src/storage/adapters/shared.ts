import { LevelUp } from 'levelup';
import { CharwiseKey } from 'charwise';
import { JsonEntry } from '../../utils/json';
import {
  AbstractBatch,
  AbstractIteratorOptions,
  AbstractLevelDOWN,
} from 'abstract-leveldown';

export type KeyType = CharwiseKey;
export type ValueType = JsonEntry;
export type BatchOperation = AbstractBatch<KeyType, ValueType>;
export type BatchOperations = BatchOperation[];
export type ReadOptions<K = KeyType> = AbstractIteratorOptions<K>;
export type KeyValue<K = KeyType, V = ValueType> = { key: K; value: V };

export class SharedAdapter {
  // @ts-ignore strictPropertyInitialization
  level: LevelUp<AbstractLevelDOWN<KeyType, ValueType>>;

  batch(operations: BatchOperations) {
    return this.level.batch(operations);
  }

  get<T = ValueType>(key: KeyType) {
    return (this.level.get(key) as unknown) as T;
  }

  query<K = KeyType, V = ValueType>(options: ReadOptions) {
    return (this.level.createReadStream(
      options
    ) as unknown) as AsyncIterableIterator<KeyValue<K, V>>;
  }

  async queryList<K = KeyType, V = ValueType>(options: ReadOptions) {
    const result: KeyValue<K, V>[] = [];
    for await (const data of this.query<K, V>(options)) {
      result.push(data);
    }
    return result;
  }
}

type SharedAdapterType = InstanceType<typeof SharedAdapter>;

export interface StorageAdapter extends SharedAdapterType {}
