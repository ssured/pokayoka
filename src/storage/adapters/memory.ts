import levelup from 'levelup';
import memdown from 'memdown';
import encode from 'encoding-down';
import charwise from 'charwise';
import { SharedAdapter, StorageAdapter, KeyType, ValueType } from './shared';

export class MemoryAdapter extends SharedAdapter implements StorageAdapter {
  constructor() {
    super();
    this.level = levelup(
      encode<KeyType, ValueType>(memdown(), {
        keyEncoding: charwise,
        valueEncoding: 'json',
      })
    );
  }
}
