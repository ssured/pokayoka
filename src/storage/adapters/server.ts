import levelup from 'levelup';
import leveldown from 'leveldown';
import encode from 'encoding-down';
import charwise from 'charwise';
import { SharedAdapter, StorageAdapter, KeyType, ValueType } from './shared';

export class ServerAdapter extends SharedAdapter implements StorageAdapter {
  constructor(public name: string) {
    super();
    this.level = levelup(
      encode<KeyType, ValueType>(leveldown(`./${name}.db`), {
        keyEncoding: charwise,
        valueEncoding: 'json',
      })
    );
  }
}
