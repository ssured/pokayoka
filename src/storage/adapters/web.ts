import levelup from 'levelup';
import levelJs from 'level-js';
import encode from 'encoding-down';
import charwise from 'charwise';
import { SharedAdapter, StorageAdapter, KeyType, ValueType } from './shared';

export class WebAdapter extends SharedAdapter implements StorageAdapter {
  constructor(public name: string) {
    super();
    this.level = levelup(
      encode<KeyType, ValueType>(levelJs(name), {
        keyEncoding: charwise,
        valueEncoding: 'json',
      })
    );
  }
}
