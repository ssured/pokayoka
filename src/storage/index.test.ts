import { Storage } from './index';
import { MemoryAdapter } from './adapters/memory';

describe('Storage', () => {
  test('it loads', () => {
    expect(Storage).toBeDefined();
    expect(MemoryAdapter).toBeDefined();
    const storage = new Storage(new MemoryAdapter());
    expect(storage).toBeDefined();
  });

  test('snapshots can be written and persisted', async () => {
    const mem = new MemoryAdapter();
    const storage = new Storage(mem);

    const obj = { id: 'test', a: 'A' };
    await storage.writeRawSnapshot(obj);

    expect(await storage.getRawSnapshot(obj.id)).toEqual(obj);
    // expect(
    //   (await mem.queryList({})).map(JSON.stringify as any).join('\n')
    // ).toEqual('');
  });

  test('snapshots are automatically merged', async () => {
    const mem = new MemoryAdapter();
    const storage = new Storage(mem);

    const obj1 = { id: 'test', a: 'A', b: 'b' };
    await storage.writeRawSnapshot(obj1);
    const contentLength1 = (await mem.queryList({})).length;

    const obj2 = { id: 'test', b: 'B' };
    await storage.writeRawSnapshot(obj2);
    const contentLength2 = (await mem.queryList({})).length;

    const result = await storage.getRawSnapshot(obj1.id);
    expect(result).toEqual({ ...obj1, ...obj2 });

    // only 2 props are stored, which means the old data is correctly removed
    expect(contentLength1).toBe(contentLength2);

    // expect(
    //   (await mem.queryList({})).map(JSON.stringify as any).join('\n')
    // ).toEqual('');
  });

  test('snapshots are automatically merged', async () => {
    const mem = new MemoryAdapter();
    const storage = new Storage(mem);

    const obj1 = { id: 'test', a: 'A', b: 'b' };
    await storage.writeRawSnapshot(obj1);
    const contentLength1 = (await mem.queryList({})).length;

    const obj2 = { id: 'test', b: 'B' };
    await storage.writeRawSnapshot(obj2);
    const contentLength2 = (await mem.queryList({})).length;

    const result = await storage.getRawSnapshot(obj1.id);
    expect(result).toEqual({ ...obj1, ...obj2 });

    // only 2 props are stored, which means the old data is correctly removed
    expect(contentLength1).toBe(contentLength2);

    // expect(
    //   (await mem.queryList({})).map(JSON.stringify as any).join('\n')
    // ).toEqual('');
  });
});
