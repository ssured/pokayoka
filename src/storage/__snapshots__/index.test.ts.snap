// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storage it stores and retrieves data 1`] = `
Array [
  Object {
    "key": Array [
      "a",
    ],
    "value": "testa",
  },
  Object {
    "key": Array [
      "b",
    ],
    "value": "testb",
  },
]
`;

exports[`Storage it stores and retrieves data 2`] = `
Array [
  Object {
    "key": Array [
      "a",
    ],
    "value": "testa",
  },
  Object {
    "key": Array [
      "b",
    ],
    "value": "testb",
  },
]
`;
